### PR TITLE
Java: make null-pointer and similar checks fatal

### PR DESCRIPTION
**Pull request by smowton**
These previously used code_assertt, which asserts but then continues -- this
changes them to use assert(condition); assume(condition), which ensures subsequent
code is not executed in conditions which are in practice impossible.

With --java-throw-runtime-exceptions this is irrelevant, as a real exception is raised
and the conditional GOTO guarding the throw has the same effect as the assume.

### CODE REVIEW

**Comment by smowton** [comment]
This will need a test-gen bump once test-gen#1794 is in.

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 1]
```diff
@@ -82,6 +83,21 @@ const std::vector<std::string> exception_needed_classes = {
   "java.lang.NegativeArraySizeException",
   "java.lang.NullPointerException"};
 
+/// Get a copy of a source location with additional assertion information
+/// \param loc: original source location
+/// \param comment: human-readable assertion description
+/// \param property_class: assertion property class
+static source_locationt with_assertion_description(
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Possibly useful, but please make this more generally available. On quick inspection almost every place that `git grep -n set_comment` highlights could benefit from this.

File: src/util/assert_or_die.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+Author: Diffblue Ltd
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_ASSERT_OR_DIE_H
+#define CPROVER_UTIL_ASSERT_OR_DIE_H
+
+#include "std_code.h"
+
+codet create_assert_or_die(const exprt &condition, const source_locationt &loc);
```
Comment (tautschnig) [inline src/util/assert_or_die.h]: 1) I think `expr_util` or `std_code` are perfectly good places to put this, no need for a new file.
2) Be specific about the return type. There is no need to hide the fact that it returns a `code_blockt`.
3) I'm not sure I like the name. How about `aborting_assert`?

File: src/java_bytecode/java_bytecode_instrument.cpp (hunk 1 above)
Comment (peterschrammel) [reply src/java_bytecode/java_bytecode_instrument.cpp]: I don't think that makes the code more readable.

**Comment by smowton** [comment]
@tautschnig @peterschrammel changes applied, please re-review.

File: src/util/std_code.cpp [hunk 3]
```diff
@@ -107,3 +107,15 @@ void code_blockt::append(const code_blockt &extra_block)
...
 }
+
+code_blockt create_fatal_assertion(
+  const exprt &condition, const source_locationt &loc)
+{
+  code_blockt result;
+  result.copy_to_operands(code_assertt(condition));
+  result.copy_to_operands(code_assumet(condition));
+  for(auto &op : result.operands())
+    op.add_source_location() = loc;
```
Comment (tautschnig) [inline src/util/std_code.cpp:118]: How about:
```
code_blockt result;
result.add(code_assertt(condition), loc);
result.add(code_assumet(condition), loc);
```
or (leaving the loop intact)
```
code_blockt result({code_assertt(condition), code_assumet(condition)});
```

**Review by tautschnig — APPROVED** [review]

**Review by martin-cs — APPROVED** [review]
Seems in line with the general state of affairs.

**Comment by smowton** [comment]
This has exposed some probably unrelated test failures. To investigate.

**Comment by smowton** [comment]
The problem was another fragile string solver test that required some luck to test; it was failing because the md5sum of the binary changed / moon wasn't aligned with Saturn, rather than due to the actual change introduced here.. I've hardened the test, hopefully now this will pass CI.

**Review by peterschrammel — APPROVED** [review]

**Comment by smowton** [comment]
Bump: https://github.com/diffblue/test-gen/pull/1830

**Comment by kroening** [comment]
BTW, related to this, I would be strongly in favour of making --java-throw-runtime-exceptions the default for jbmc. The Java semantics is very clear on these, this isn't the same situation as in C.

There can be a --java-runtime-exceptions-block or the like.

**Review by kroening — APPROVED** [review]

**Comment by smowton** [comment]
@kroening as a prereq to --java-throw-runtime-exceptions we need to get a lot better at modelling control flow, otherwise a lot of implausible exceptional flows blur symex's certainty about the state of the world and the formula grows remarkably.

**Comment by peterschrammel** [comment]
@smowton, could using https://github.com/diffblue/cbmc/blob/develop/src/analyses/goto_check.cpp#L1478 improve the situation (at least for NPEs)?

**Pull request merged**